### ISSUE TITLE
Type inference in lambda-like expressions

### ISSUE THREAD

**Issue by SergioMorchon** [palantir/tslint#333]
If I write the following line:

``` typescript
"Hello world".split(/\s/).forEach(word => console.log(word));
```
tslint says ```expected parameter: 'word' to have a typedef```
But as the new 1.4 TS version, 'word' is a string variable.

**Comment by SergioMorchon** [comment]
It would be great to be able to distinguish, by the ```typedef``` parameter, the next situations:
```"parameter"``` checks type specifier of function parameters
```"inline-function-parameter"``` checks type specifier of inline functions (i.e. forEach callbacks) or relays on type inference.

Because when I declare a function that will be used afeter, I want to declare its parameter types, but when I declare a callback function I want the compiler to infer the type and check it, without possibly cast the parameter values to erroneous types.

**Comment by adidahiya** [comment]
I like this idea; it'd be nice to split out the `typdef:parameter` option into:

1. `typedef:parameter`, which checks `function` call signatures & class methods
2. `typedef:arrow` which checks arrow `=>` call signatures

thoughts on the naming / delineation here @gscshoyru?

**Comment by gscshoyru** [comment]
So to be clear -- ```typedef:arrow``` covers both ```param => ``` as well as ```(param1, param2) =>``` right? 

Yeah, that makes sense as a thing to split up.

I'm not sure calling it "parameter" makes as much sense, though, since both have parameters -- maybe functionParameter and arrowParameter? That might be a bit too verbose, though.

**Comment by JKillian** [comment]
Closed by #965 - there's now an `arrow-parameter` option that will enable checking for typedefs on arrow functions, where as the current `parameter` option will enable checking for non-arrow functions.
